nft_ingester/src/program_transformers/bubblegum/db.rs

```diff
@@ -114,42 +111,157 @@ where
 ) -> Result<(), IngesterError>
 where
     T: ConnectionTrait + TransactionTrait,
 {
-    let update_one = if let Some(seq) = seq {
-        asset::Entity::update(model).filter(
-            Condition::all()
-                .add(asset::Column::Id.eq(id.clone()))
-                .add(asset::Column::Seq.lte(seq)),
-        )
-    } else {
-        asset::Entity::update(model).filter(asset::Column::Id.eq(id.clone()))
+    let model = asset::ActiveModel {
+        id: Set(id),
+        leaf: Set(leaf),
+        leaf_seq: Set(seq),
+        ..Default::default()
     };
 
-    match update_one.exec(txn).await {
-        Ok(_) => Ok(()),
-        Err(err) => match err {
-            DbErr::RecordNotFound(ref s) => {
-                if s.contains("None of the database rows are affected") {
-                    warn!(
-                        "Update failed. No asset found for id {}.",
-                        bs58::encode(id).into_string()
-                    );
-                    Ok(())
-                } else {
-                    Err(IngesterError::from(err))
-                }
-            }
-            _ => Err(IngesterError::from(err)),
-        },
+    let mut query = asset::Entity::insert(model)
+        .on_conflict(
+            OnConflict::column(asset::Column::Id)
+                .update_columns([asset::Column::Leaf, asset::Column::LeafSeq])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+
+    // If we are indexing decompression we will update the leaf regardless of if we have previously
+    // indexed decompression and regardless of seq.
+    if !was_decompressed {
+        query.sql = format!(
+            "{} WHERE (NOT asset.was_decompressed) AND (excluded.leaf_seq > asset.leaf_seq OR asset.leaf_seq IS NULL)",
+            query.sql
+        );
     }
+
+    txn.execute(query)
+        .await
+        .map_err(|db_err| IngesterError::StorageWriteError(db_err.to_string()))?;
+
+    Ok(())
+}
+
+pub async fn upsert_asset_with_owner_and_delegate_info<T>(
+    txn: &T,
+    id: Vec<u8>,
+    owner: Vec<u8>,
+    delegate: Option<Vec<u8>>,
+    seq: i64,
+) -> Result<(), IngesterError>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    let model = asset::ActiveModel {
+        id: Set(id),
+        owner: Set(Some(owner)),
+        delegate: Set(delegate),
+        owner_delegate_seq: Set(Some(seq)), // gummyroll seq
+        ..Default::default()
+    };
+
+    let mut query = asset::Entity::insert(model)
+        .on_conflict(
+            OnConflict::column(asset::Column::Id)
+                .update_columns([
+                    asset::Column::Owner,
+                    asset::Column::Delegate,
+                    asset::Column::OwnerDelegateSeq,
+                ])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+    query.sql = format!(
+        "{} WHERE excluded.owner_delegate_seq > asset.owner_delegate_seq OR asset.owner_delegate_seq IS NULL",
```

Comment:
delegate_seq is only null at first right? never later, right?

Reply:
Correct, it is only null when when the asset row is created by some other upsert that does not put a value there. And yes it should never go back to null after its been populated.